src/Petri.jl

```diff
@@ -0,0 +1,101 @@
+module Petri
+
+using Catlab
+using Catlab.CategoricalAlgebra.CSets
+using Catlab.Present
+using Catlab.Theories
+
+# Petri Nets
+############
+
+@present TheoryPetri(FreeCategory) begin
+    T::Ob
+    S::Ob
+    I::Ob
+    O::Ob
+
+    it::Hom(I,T)
+    is::Hom(I,S)
+    ot::Hom(O,T)
+    os::Hom(O,S)
+end
+
+const AbstractPetri = AbstractCSetType(TheoryPetri)
+const Petri = CSetType(TheoryPetri, index=[:is,:it,:os,:ot])
+
+ns(p::AbstractPetri) = nparts(p,:S)
+nt(p::AbstractPetri) = nparts(p,:T)
+ni(p::AbstractPetri) = nparts(p,:I)
+no(p::AbstractPetri) = nparts(p,:O)
+
+add_species!(p::AbstractPetri) = add_part!(p,:S)
+add_species!(p::AbstractPetri,n) = add_parts!(p,:S,n)
+
+add_transition!(p::AbstractPetri) = add_part!(p,:T)
+add_transitions!(p::AbstractPetri,n) = add_parts!(p,:T,n)
+
+add_input!(p::AbstractPetri,t,s) = add_part!(p,:I,(is=s,it=t))
+add_inputs!(p::AbstractPetri,n,t,s) = add_parts!(p,:I,n,(is=s,it=t))
+
+add_output!(p::AbstractPetri,t,s) = add_part!(p,:O,(os=s,ot=t))
+add_output!(p::AbstractPetri,n,t,s) = add_parts!(p,:O,n,(os=s,ot=t))
+
+# Note: although indexing makes this pretty fast, it is often faster to bulk-convert
+# the Petri net into a transition matrix, if you are working with all of the transitions
+inputs(p::AbstractPetri,t) = subpart(p,incident(p,t,:it),:is)
+outputs(p::AbstractPetri,t) = subpart(p,incident(p,t,:ot),os)
+
+totransmat(p::AbstractPetri) = begin
```

Comment:
Would it be more idiomatic to call this function `transition_matrix` or make it a constructor of `Array{Int,2}`?

Reply:
Do you mean `Array{Int,3}`? Also, I don't know what would be more idiomatic, and I am not attached to nomenclature; whatever you think is the right way to do name this I will do.

Reply:
Oh yeah, 3D array makes sense, I got confused by the two in this call `A = zeros(Int64,(nt(p),ns(p),2))` but that is saying the third dimension has length 2. Looking at the code below where this gets used, I think that making this two sparse matrices one for `A[:,:,1]` and the other for `A[:,:,2]` and returning the pair fo them as Tuple{CSCMatrix{Int}, CSCMatrix{Int}}` would be better.